src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,447 @@
+  // from the input is after _content_end.  The difference between
+  // these two pointers is constant, except when characters change
+  // from being in the current line to being (logically) before it,
+  // when next is called.
+
+  bool is_sane() const {
+    assert((_buffer == nullptr) == (_buffer_size == 0), "");
+    assert(_content_end <= _buffer_size, "");
+    assert(_beg <= _end && _end <= _content_end, "");
+    assert(_end <= _next && _next <= _content_end + NEXT_PHANTOM, "");
+    assert(_buffer_size == 0 || _next <= _buffer_size, "");
+    return true;
+  }
+
+  bool need_to_read() const {
+    assert(is_sane(), "");
+    return _next == _end;
+  }
+  bool have_current_line() const {
+    assert(is_sane(), "");
+    // _beg < _content_end because there is an \0 (was \n) at _end,
+    // or else it is a non-empty partial line and the \0 is at
+    // _content_end.  In either case, if _end == _next we are
+    // still searching for more input.
+    return (_beg < _content_end && _end < _next);
+  }
+  bool definitely_done() const {
+    assert(is_sane(), "");
+    // If _beg < _content_end we still have a line of some sort.
+    // Otherwise, if _next > _content_end, we have seen EOF or error.
+    return (_beg == _content_end && _next > _content_end);
+  }
+
+  // Reset indexes within the buffer to point to no content.
+  void clear_buffer();
+
+  // Reset indexes within the buffer to point to the given content.
+  // This is where we scan for newlines as well.
+  void set_buffer_content(size_t content_start, size_t content_end);
+
+  // Try to make the buffer bigger.  This may be necessary in order to
+  // buffer a very long line.  Returns false if there was an
+  // allocation failure.
+  //
+  // On allocation failure, just make do with whatever buffer there
+  // was to start with; the caller must check for this condition and
+  // avoid buffering more data in the non-expanded buffer.  However,
+  // the buffer will always be non-null, so at least one line can be
+  // buffered, if it is of normal size.
+  bool expand_buffer(size_t new_length);
+
+  // Make sure there is at least one line in the buffer, and set
+  // _beg/_end to indicate where it is.  Any content before _beg can
+  // be overwritten to make more room in the buffer.  If there is no
+  // more input, set the state up to indicate we are done.
+  bool fill_buffer();
+
+  // Find some room in the buffer so we call read on it.
+  // This might call expand_buffer but will try not to.
+  // The assumption is that read already buffers slow I/O calls.
+  // The purpose for the small buffer managed here is to store whole lines,
+  // and perhaps edit them in-place.
+  void prepare_to_fill_buffer(size_t& fill_offset, size_t& fill_length);
+
+  // Quick check for an initially incomplete buffer...
+  void preload() const {
+    if (need_to_read()) {
+      const_cast<inputStream*>(this)->fill_buffer();
+    }
+  }
```

Comment:
Why `const_cast` and assign this method `const` when it clearly is not? This shouldn't be `const`, is my point.

Reply:
The method is const because the logical state of the stream is invariant, as visible to the API user.  If the implementation needs an invisible internal state change, it needs a const-cast (or mutable field, sometimes, will work).  If `preload` were made non-const as you suggest, we’d need to move the const-casting elsewhere, and it would be less clear that `preload` preserves API-visible state.  So the code, as it is, is the most convenient place to put the const-cast, as an internal implementation decision.

Reply:
This page suggests using the `mutable` keyword for this situation (but `const_cast<>` is acceptable as a last resort)
https://isocpp.org/wiki/faq/const-correctness#mutable-data-members

However, to use `mutable` in this case, we would probably need to declare every field as `mutable`, this means that the meaning of `const` will be very difficult to understand (at least we can't use the C++ compiler to tell us which function can't (should) be `const` and which function cannot be `const`.)

Reply:
On a separate note, `preload()` can cause data to be read from the input. For non-seekable input (such as sockets), this doesn't seem like a `const` operation to me.

Reply:
I could be wrong about this, but it seems like it’s completely up to the C++ class to define what aspects of its implementation are fixed for its `const` methods.  The `const` keyword does not mean “everything is immutable about this, at all levels and inside all boundaries”.  But then it’s up to the class author to choose which boundary `const` applies to.  The thing wrapped inside the i-stream is, I will claim, implementation which can change state even in `const` functions.

If not, the alternative is to have almost no `const` functions at all in the i-stream class.  That’s less useful, because `const` means something very useful, in the context of the i-stream class.  It means that the line will not shift.  You can keep on reading, and even writing, the line buffer, as long as you call only `const` functions.

The reuse of the line buffer, in this way, is a core part of the i-stream design, and part of its value proposition:  You don’t get a new allocation on every read-line op (as you would with Java).  But in order to draw bright lines around THAT very useful notion of invariance (during which line buffers are reusable), you NEED `preload` to be const, even if it does internal book keeping, even if it consults the input source at times.

In short, a rigid idea of `const` is inconsistent with the carefully balanced performance characteristics of this design.

Reply:
I had to read up on this and I believe that John is right.

To quote https://isocpp.org/wiki/faq/const-correctness#const-member-fns

>The trailing const on inspect() member function should be used to mean the method won’t change the object’s abstract (client-visible) state. That is slightly different from saying the method won’t change the “raw bits” of the object’s struct.

I'm not objecting to the usage of `const` for this method, considering this. Reviewing code is learning on the job!